archaius2-core/src/main/java/com/netflix/archaius/ConfigProxyFactory.java

```diff
@@ -182,131 +180,145 @@ interface MethodInvoker<T> {
         
         final String prefix = derivePrefix(annot, initialPrefix);
-        
+
+        // There's a circular dependency between these maps and the proxy object. They must be created first because the
+        // proxy's invocation handler needs to keep a reference to them, but the proxy must be created before they get
+        // filled because we may need to call methods on the interface in order to fill the maps :-|
+        final Map<Method, PropertyValueGetter<?>> invokers = new HashMap<>();
+        final Map<Method, String> propertyNames = new HashMap<>();
+
+        final InvocationHandler handler = new ConfigProxyInvocationHandler<>(type, prefix, invokers, propertyNames);
+
+        final T proxyObject = (T) Proxy.newProxyInstance(type.getClassLoader(), new Class[] { type }, handler);
+
         // Iterate through all declared methods of the class looking for setter methods.
         // Each setter will be mapped to a Property<T> for the property name:
         //      prefix + lowerCamelCaseDerivedPropertyName
-        final Map<Method, MethodInvoker<?>> invokers = new HashMap<>();
-        final Map<Method, String> propertyNames = new HashMap<>();
+        for (Method method : type.getMethods()) {
+            MethodInvokerHolder methodInvokerHolder = buildInvokerForMethod(type, prefix, method, proxyObject, immutable);
 
-        final InvocationHandler handler = (proxy, method, args) -> {
-            MethodInvoker<?> invoker = invokers.get(method);
-            if (invoker != null) {
-                return invoker.invoke(args);
-            }
-            if ("equals".equals(method.getName())) {
-            	return proxy == args[0];
-            }
-            else if ("hashCode".equals(method.getName())) {
-            	return System.identityHashCode(proxy);
-            }
-            else if ("toString".equals(method.getName())) {
-                StringBuilder sb = new StringBuilder();
-                sb.append(type.getSimpleName()).append("[");
-                sb.append(invokers.entrySet().stream().map(entry -> {
-                	StringBuilder sbProperty = new StringBuilder();
-                	sbProperty.append(propertyNames.get(entry.getKey()).substring(prefix.length())).append("='");
-                    try {
-                    	sbProperty.append(entry.getValue().invoke(null));
-                    } catch (Exception e) {
-                    	sbProperty.append(e.getMessage());
-                    }
-                    sbProperty.append("'");
-                    return sbProperty.toString();
-                }).collect(Collectors.joining(",")));
-                sb.append("]");
-                return sb.toString();
+            propertyNames.put(method, methodInvokerHolder.propertyName);
+
+            if (immutable) {
+                // Cache the current value of the property and always return that.
+                // Note that this will fail for parametrized properties!
```

Comment:
There are a few instances where parameterized is misspelled parametrized

Reply:
Oh it looks like both are valid, but I think it'd be best if we stuck with one or the other regardless

Reply:
LOL, thanks for catching this!